pkg/operators/ebpf/tracer.go

```diff
@@ -0,0 +1,243 @@
+	if traceMap.Type != ebpf.RingBuf && traceMap.Type != ebpf.PerfEventArray {
+		return fmt.Errorf("map %q has a wrong type, expected: ringbuf or perf event array, got: %s",
+			traceMap.Name, traceMap.Type.String())
+	}
+	return nil
+}
+
+func (i *ebpfInstance) populateTracer(t btf.Type, varName string) error {
+	i.logger.Debugf("populating tracer %q", varName)
+
+	parts := strings.Split(varName, typeSplitter)
+	if len(parts) != 3 {
+		return fmt.Errorf("invalid tracer info: %q", varName)
+	}
+
+	name := parts[0]
+	mapName := parts[1]
+	structName := parts[2]
+
+	i.logger.Debugf("> name       : %q", name)
+	i.logger.Debugf("> map name   : %q", mapName)
+	i.logger.Debugf("> struct name: %q", structName)
+
+	tracerConfig := i.config.Sub("tracers." + name)
+	if tracerConfig != nil {
+		if configMapName := tracerConfig.GetString("mapName"); configMapName != "" && configMapName != mapName {
+			return fmt.Errorf("validating tracer %q: mapName %q in eBPF program does not match %q from metadata file",
+				name, configMapName, mapName)
+		}
+		if configStructName := tracerConfig.GetString("structName"); configStructName != "" && configStructName != structName {
+			return fmt.Errorf("validating tracer %q: structName %q in eBPF program does not match %q from metadata file",
+				name, configStructName, structName)
+		}
+		i.logger.Debugf("> successfully validated with metadata")
+	}
+
+	if _, ok := i.tracers[name]; ok {
+		i.logger.Debugf("tracer %q already defined, skipping", name)
+		return nil
+	}
+
+	tracerMap, ok := i.collectionSpec.Maps[mapName]
+	if !ok {
+		return fmt.Errorf("map %q not found in eBPF object", mapName)
+	}
+
+	if err := validateTracerMap(tracerMap); err != nil {
+		return fmt.Errorf("trace map is invalid: %w", err)
+	}
+
+	var btfStruct *btf.Struct
+	if err := i.collectionSpec.Types.TypeByName(structName, &btfStruct); err != nil {
+		return fmt.Errorf("finding struct %q in eBPF object: %w", structName, err)
+	}
+
+	i.logger.Debugf("adding tracer %q", name)
+	i.tracers[name] = &Tracer{
+		Tracer: metadatav1.Tracer{
+			MapName:    mapName,
+			StructName: btfStruct.Name,
+		},
+		eventSize: btfStruct.Size,
+	}
+
+	err := i.populateStructDirect(btfStruct)
+	if err != nil {
+		return fmt.Errorf("populating struct %q for tracer %q: %w", btfStruct.Name, name, err)
+	}
+
+	return nil
+}
+
+func (t *Tracer) receiveEvents() error {
+	switch t.mapType {
+	case ebpf.RingBuf:
+		return t.receiveEventsFromRingReader()
+	case ebpf.PerfEventArray:
+		return t.receiveEventsFromPerfReader()
+	default:
+		return fmt.Errorf("invalid map type")
+	}
+}
+
+func (t *Tracer) receiveEventsFromRingReader() error {
+	slowBuf := make([]byte, t.eventSize)
+	lastSlowLen := 0
+	for {
+		rec, err := t.ringbufReader.Read()
+		if err != nil {
+			return err
+		}
+		data := t.ds.NewData()
+		sample := rec.RawSample
+		if uint32(len(rec.RawSample)) < t.eventSize {
+			// event is truncated; we need to copy
+			copy(slowBuf, rec.RawSample)
```

Comment:
> I think there is code for this upstream.

Could you please point me into that direction?

> Also, what about having event with dynamic size?

That would need much more careful handling in other places (validating fields before accessing members). We can optimize that later, but I think for now this would add quite a bit of complexity.